### PR TITLE
Prevent collision of imported type with exported declarations in current module

### LINKED ISSUE #31031: Importing a type with same name as an export
**TypeScript Version:**  3.5.0-dev.20190419

**Search Terms:** import export conflict priority

**Code**

```ts
/// foo.ts
export interface Foo { a: number; }

/// test.ts
import { Foo } from './foo'
export interface Foo { b: number; }
let foo: Foo = { a: 2 }
```

**Expected behavior:** Declaration conflict on `test.ts` (imported Foo vs exported Foo).

**Actual behavior:** No error. Instead, (see last line) `Foo` refers to the imported Foo, and the exported Foo becomes unaccessible. Is this expected?

If `export interface` is changed to `export class`, then a declaration conflict appears.  
If `export` is removed in initial case, a declaration conflict also appears.

**Playground Link:** I don't know how to reproduce this with one file.

**Related Issues:** I searched issues, stackoverflow and docs (declaration merging, imports and exports, namespaces) but couldn't find any references to this behaviour.

### PR DESCRIPTION
**Pull request by collin5**
Fixes #31031

### CODE REVIEW

**Comment by collin5** [comment]
@andrewbranch rebased. Thank you.

**Comment by andrewbranch** [comment]
@typescript-bot user test this

**Comment by andrewbranch** [comment]
User tests only turn up failures in VS Code: https://github.com/collin5/TypeScript/pull/1/files#diff-0e62d568d28b82cee9e66b2b8fafd290R7

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -30089,6 +30089,7 @@ namespace ts {
                 // Based on symbol.flags we can compute a set of excluded meanings (meaning that resolved alias should not have,
                 // otherwise it will conflict with some local declaration). Note that in addition to normal flags we include matching SymbolFlags.Export*
                 // in order to prevent collisions with declarations that were exported from the current module (they still contribute to local names).
+                symbol = getMergedSymbol(symbol.exportSymbol || symbol);
```
Comment (andrewbranch) [inline src/compiler/checker.ts]: Just to fill in some historical context here: I was curious why this was needed when #7583 and #7591 only deal with the flags on the local symbol. Dug around a bit and found #16766. If you look at the changes in the binder there, information about the export _used_ to be attached to the local symbol when it was exporting a type, namespace, or value, but #16766 removed that info for types and namespaces. So, the info we need for this function here used to exist on the local symbol, but now we have to go to the export symbol to get it. This implementation appears correct (although I wonder if #16766 would have been done differently if the flags had been used here).

**Comment by Jessidhia** [comment]
I hate to be the https://xkcd.com/1172/ person but this is actually something I liberally use to work around `--isolatedModules`-related limitations when re-exporting types under the same name 🤣 

```ts
// a.ts
export interface A {}

// index.ts
import { A } from './a'
export type A = A
```

No big deal, it'll just take some, uh, intensive codemodding.

---

The best way of dealing with this and correctly be able to elide type-only reexports with `--isolatedModules` / `@babel/preset-typescript` would probably be to have an `export { type A } from './a'` syntax, à la Flow.

**Review by Wiseboy11 — COMMENTED** [review]
Congrats

**Pull request merged**